File: ruby/lib/steam/packets/rcon/rcon_packet.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'steam/packets/steam_packet'

# This module is included by all classes representing a packet used by Source's
# RCON protocol
#
# It provides a basic implementation for initializing and serializing such a
# packet.
#
# @author Sebastian Staudt
# @see RCONPacketFactory
module RCONPacket

  include SteamPacket

  # Header for authentication requests
  SERVERDATA_AUTH = 3

  # Header for replies to authentication attempts
  SERVERDATA_AUTH_RESPONSE = 2

  # Header for command execution requests
  SERVERDATA_EXECCOMMAND = 2

  # Header for packets with the output of a command execution
  SERVERDATA_RESPONSE_VALUE = 0

  # Returns the request ID used to identify the RCON communication
  #
  # @return [Fixnum] The request ID used to identify the RCON communication
  attr_reader :request_id

  # Creates a new RCON packet object with the given request ID, type and
  # content data
  #
  # @param [Fixnum] request_id The request ID for the current RCON
  #        communication
  # @param [Fixnum] rcon_header The header for the packet type
  # @param [String] rcon_data The raw packet data
  def initialize(request_id, rcon_header, rcon_data)
    super rcon_header, "#{rcon_data}\0\0"

    @request_id = request_id
  end

  # Returns the raw data representing this packet
  #
  # @return [String] A string containing the raw data of this RCON packet
  def to_s
    [@content_data.length + 8, @request_id, @header_data, @content_data.string].pack('V3a*')
  end

end

File: php/lib/steam/packets/rcon/RCONPacket.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2009, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage RCONPacket
 */

require_once STEAM_CONDENSER_PATH . 'steam/packets/SteamPacket.php';

abstract class RCONPacket extends SteamPacket
{
	const SERVERDATA_AUTH = 3;
	const SERVERDATA_AUTH_RESPONSE = 2;
	const SERVERDATA_EXECCOMMAND = 2;
	const SERVERDATA_RESPONSE_VALUE = 0;

	/**
	 * @var long
	 */
	private $requestId;

	public function __construct($requestId, $rconHeader, $rconData = null)
	{
		parent::__construct($rconHeader, "$rconData\0\0");

		$this->requestId = $requestId;
	}

	public function getBytes()
	{
		$contentData = $this->contentData->_array();
		return pack("V3a*", strlen($contentData) + 8, $this->requestId, $this->headerData, $contentData);
	}

	public function getRequestId()
	{
		return $this->requestId;
	}
}
?>
